torchmetrics/functional/text/squad.py

```diff
@@ -0,0 +1,292 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Adapted from:
+# Link: https://worksheets.codalab.org/rest/bundles/0x6b567e1cf2e041ec80d7098f031c5c9e/contents/blob/
+# Link: https://github.com/huggingface/datasets/blob/master/metrics/squad/squad.py
+import re
+import string
+from collections import Counter
+from typing import Dict, List, Tuple, Union
+
+import torch
+from torch import Tensor, tensor
+
+from torchmetrics.utilities import rank_zero_warn
+
+SINGLE_PRED_TYPE = Dict[str, str]
+PREDS_TYPE = Union[SINGLE_PRED_TYPE, List[SINGLE_PRED_TYPE]]
+SINGLE_TARGET_TYPE = Dict[str, Union[str, Dict[str, Union[List[str], List[int]]]]]
+TARGETS_TYPE = Union[SINGLE_TARGET_TYPE, List[SINGLE_TARGET_TYPE]]
+
+
+def normalize_text(s: str) -> str:
+    """Lower text and remove punctuation, articles and extra whitespace."""
+
+    def remove_articles(text: str) -> str:
+        return re.sub(r"\b(a|an|the)\b", " ", text)
+
+    def white_space_fix(text: str) -> str:
+        return " ".join(text.split())
+
+    def remove_punc(text: str) -> str:
+        exclude = set(string.punctuation)
+        return "".join(ch for ch in text if ch not in exclude)
+
+    def lower(text: str) -> str:
+        return text.lower()
+
+    return white_space_fix(remove_articles(remove_punc(lower(s))))
+
+
+def get_tokens(s: str) -> List[str]:
+    """Split a sentence into separate tokens."""
+    if not s:
+        return []
+    return normalize_text(s).split()
+
+
+def compute_f1_score(predictied_answer, target_answer) -> Tensor:
+    """Compute F1 Score for two sentences."""
+    target_tokens: Tensor = get_tokens(target_answer)
+    predicted_tokens: Tensor = get_tokens(predictied_answer)
+    common = Counter(target_tokens) & Counter(predicted_tokens)
+    num_same: Tensor = tensor(sum(common.values()))
+    if len(target_tokens) == 0 or len(predicted_tokens) == 0:
+        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
+        return tensor(int(target_tokens == predicted_tokens))
+    if num_same == 0:
+        return tensor(0.0)
+    precision: Tensor = 1.0 * num_same / tensor(len(predicted_tokens))
+    recall: Tensor = 1.0 * num_same / tensor(len(target_tokens))
+    f1: Tensor = (2 * precision * recall) / (precision + recall)
```

Comment:
I think explicit typing `mypy` doesn't require explicit typing here, does it?

Reply:
Regarding this, I generally try to type the variables as much as possible.
Helps in code reading and understanding and also IDE typing.

Reply:
The default convention in torchmetrics is that if mypy is able to infer the type inside a method/function then the variable should not be typed.